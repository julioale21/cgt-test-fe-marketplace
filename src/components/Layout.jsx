import { Outlet } from 'react-router-dom';
import { Header } from '../components/shared/Header';
import { Box } from '@mui/material';
import { useCart } from '../hooks/useCart';

const Layout = () => {
  const { items } = useCart();

  const cartItemsCount = items.reduce((total, item) => total + item.quantity, 0);

  return (
    <Box sx={{ display: 'flex', flexDirection: 'column', minHeight: '100vh' }}>
      <Header cartItemsCount={cartItemsCount} />

      <Box component="main" sx={{ flexGrow: 1, p: 3 }}>
        <Outlet />
      </Box>

      <Box component="footer" sx={{ p: 2, textAlign: 'center', mt: 'auto' }}>
        <p>Â© 2025</p>
      </Box>
    </Box>
  );
};

export { Layout };
