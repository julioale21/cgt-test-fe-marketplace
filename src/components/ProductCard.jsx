import React from 'react';

import { Link } from 'react-router-dom';

import { Button, Card, CardActions, CardContent, CardMedia, Grid, Typography } from '@mui/material';
import ShoppingCartIcon from '@mui/icons-material/ShoppingCart';

const ProductCard = ({ product }) => {
  return (
    <Grid item xs={12} sm={6} md={4} lg={3}>
      <Card sx={{ height: '100%', display: 'flex', flexDirection: 'column' }}>
        <CardMedia
          component="img"
          height="200"
          image={product.image}
          alt={product.name}
          sx={{
            objectFit: 'cover',
            height: '200px'
          }}
        />
        <CardContent sx={{ flexGrow: 1 }}>
          <Typography gutterBottom variant="h6" component="h2">
            {product.name}
          </Typography>
          <Typography variant="body2" color="text.secondary" sx={{ mb: 2 }}>
            {product.description}
          </Typography>
          <Typography variant="h6" color="primary" sx={{ mt: 2 }}>
            ${product.price.toFixed(2)}
          </Typography>
          <Typography variant="body2" color="text.secondary" sx={{ mt: 1 }}>
            Formats: {product.format.join(', ')}
          </Typography>
        </CardContent>
        <CardActions sx={{ justifyContent: 'space-between' }}>
          <Button component={Link} to={`/products/${product.id}`} size="small" color="primary">
            View Details
          </Button>
          <Button size="small" variant="contained" color="primary" startIcon={<ShoppingCartIcon />}>
            Add to Cart
          </Button>
        </CardActions>
      </Card>
    </Grid>
  );
};

export { ProductCard };
