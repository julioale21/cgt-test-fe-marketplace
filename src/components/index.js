export { CartHeader } from "./cart/CartHeader";
export { CartItem } from "./cart/CartItem";
export { CartItemList } from "./cart/CartItemList";
export { CartSummary } from "./cart/CartSummary";
export { CartSummaryRow } from "./cart/CartSummaryRow";
export { ContainedButton } from "./buttons/ContainedButton";
export { CustomAlert } from "./shared/CustomAlert";
export { EmptyCart } from "./cart/EmptyCart";
export { Header } from "./shared/Header";
export { Layout } from "./Layout";
export { OutlinedButton } from "./buttons/OutlinedButton";
export { PaymentMethodIcon } from "./cart/PaymentMethodIcon";
export { ProductCard } from "./product/ProductCard";
export { ProductDetails } from "./product/ProductDetail";
export { ProductImage } from "./product/ProductImage";
export { ProductList } from "./product/ProductList";
export { ProductSpecs } from "./product/ProductSpecs";
export {ProductHeader} from "./product/ProductHeader";
