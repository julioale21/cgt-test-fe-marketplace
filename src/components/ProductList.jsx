import React from 'react';
import { Grid } from '@mui/material';
import { ProductCard } from './ProductCard';

const ProductList = ({ products }) => {
  return (
    <Grid container spacing={4}>
      {products.map((product) => (
        <ProductCard key={product.id} product={product} />
      ))}
    </Grid>
  );
};

export { ProductList };
