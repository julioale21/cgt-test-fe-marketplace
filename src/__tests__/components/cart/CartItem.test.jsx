import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { CartItem } from '../../../components';

describe('CartItem', () => {
  const mockItem = {
    id: 1,
    name: '3D Mountain Model',
    price: 29.99,
    image: '/test-image.jpg',
    quantity: 2
  };

  const defaultProps = {
    item: mockItem,
    onIncrement: jest.fn(),
    onDecrement: jest.fn(),
    onRemove: jest.fn()
  };

  it('matches snapshot', () => {
    const { container } = render(<CartItem {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it('displays item details correctly', () => {
    render(<CartItem {...defaultProps} />);

    expect(screen.getByText(mockItem.name)).toBeInTheDocument();
    expect(screen.getByText(`$${mockItem.price.toFixed(2)}`)).toBeInTheDocument();
    expect(screen.getByText(mockItem.quantity.toString())).toBeInTheDocument();

    const image = screen.getByRole('img');
    expect(image).toHaveAttribute('src', mockItem.image);
    expect(image).toHaveAttribute('alt', mockItem.name);
  });

  it('calls onIncrement when plus button is clicked', async () => {
    render(<CartItem {...defaultProps} />);
    const user = userEvent.setup();

    await user.click(screen.getByTestId('AddIcon').parentElement);
    expect(defaultProps.onIncrement).toHaveBeenCalledTimes(1);
  });

  it('calls onDecrement when minus button is clicked', async () => {
    render(<CartItem {...defaultProps} />);
    const user = userEvent.setup();

    await user.click(screen.getByTestId('RemoveIcon').parentElement);
    expect(defaultProps.onDecrement).toHaveBeenCalledTimes(1);
  });

  it('calls onRemove when delete button is clicked', async () => {
    render(<CartItem {...defaultProps} />);
    const user = userEvent.setup();

    await user.click(screen.getByTestId('DeleteIcon').parentElement);
    expect(defaultProps.onRemove).toHaveBeenCalledTimes(1);
  });

  it('applies correct styles to the image', () => {
    render(<CartItem {...defaultProps} />);
    const image = screen.getByRole('img');

    expect(image).toHaveStyle({
      width: '80px',
      height: '80px',
      objectFit: 'cover'
    });
  });

  it('applies small size to all icon buttons', () => {
    render(<CartItem {...defaultProps} />);
    const buttons = screen.getAllByRole('button');

    buttons.forEach((button) => {
      expect(button).toHaveClass('MuiIconButton-sizeSmall');
    });
  });

  it('applies error color to delete button', () => {
    render(<CartItem {...defaultProps} />);
    const deleteButton = screen.getByTestId('DeleteIcon').parentElement;

    expect(deleteButton).toHaveClass('MuiIconButton-colorError');
  });

  it('centers quantity display', () => {
    render(<CartItem {...defaultProps} />);
    const quantityElement = screen.getByText(mockItem.quantity.toString());

    expect(quantityElement).toHaveStyle({
      textAlign: 'center',
      minWidth: '32px'
    });
  });

  it('formats price with 2 decimal places', () => {
    const itemWithLongPrice = {
      ...mockItem,
      price: 29.9999
    };

    render(<CartItem {...defaultProps} item={itemWithLongPrice} />);
    expect(screen.getByText('$30.00')).toBeInTheDocument();
  });

  it('handles missing image gracefully', () => {
    const itemWithoutImage = {
      ...mockItem,
      image: ''
    };

    render(<CartItem {...defaultProps} item={itemWithoutImage} />);
    const image = screen.getByRole('img');
    expect(image).toHaveAttribute('src', '');
    expect(image).toBeInTheDocument();
  });
});
