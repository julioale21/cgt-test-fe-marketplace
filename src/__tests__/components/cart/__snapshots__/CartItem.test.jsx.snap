// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem matches snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1fdopfm-MuiPaper-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
  >
    <div
      class="MuiBox-root css-j0ozid"
    >
      <img
        alt="3D Mountain Model"
        class="MuiBox-root css-npxvs5"
        src="/test-image.jpg"
      />
      <div
        class="MuiBox-root css-i9gxme"
      >
        <h6
          class="MuiTypography-root MuiTypography-subtitle1 css-1i3tsf3-MuiTypography-root"
        >
          3D Mountain Model
        </h6>
        <p
          class="MuiTypography-root MuiTypography-body2 css-1wle3ir-MuiTypography-root"
        >
          $
          29.99
        </p>
      </div>
      <div
        class="MuiBox-root css-axw7ok"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-155wbr6-MuiButtonBase-root-MuiIconButton-root"
          id=":r0:"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-loadingIndicator css-165vu30-MuiIconButton-loadingIndicator"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="RemoveIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13H5v-2h14z"
            />
          </svg>
        </button>
        <p
          class="MuiTypography-root MuiTypography-body1 css-owqrw6-MuiTypography-root"
        >
          2
        </p>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-155wbr6-MuiButtonBase-root-MuiIconButton-root"
          id=":r1:"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-loadingIndicator css-165vu30-MuiIconButton-loadingIndicator"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="AddIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6z"
            />
          </svg>
        </button>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorError MuiIconButton-sizeSmall css-195ve3w-MuiButtonBase-root-MuiIconButton-root"
          id=":r2:"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-loadingIndicator css-165vu30-MuiIconButton-loadingIndicator"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="DeleteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
